### DISCUSSION TITLE
[Question]: Upload file fail with timeout

### DISCUSSION THREAD

**Discussion by JamieWuJinYi** [danny-avila/LibreChat#2634] [Troubleshooting]
### What is your question?

I uploaded a 2.5MB text, and the interface reported a timeout error after waiting for about 1.1 minutes.

After some time this file appeared in my attachments
![image](https://github.com/danny-avila/LibreChat/assets/100826652/2877ca1b-05fe-404f-ab01-ee3201e4c5c5)

I think this should be related to the timeout setting, but I have already set the timeout in librechat.yaml. Is there any other setting I can make?

### More Details

Here's my working setup with the librechat.yaml file:
![image](https://github.com/danny-avila/LibreChat/assets/100826652/63f7bb5b-6895-4490-92e2-40d1f2d629fa)

### Screenshots

![image](https://github.com/danny-avila/LibreChat/assets/100826652/fa8af318-7d9c-4647-a6b1-5dbd9044cd40)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
since you're uploading via OpenAI endpoint, it's likely timing out between the rag_api and main librechat api.

is the rag api reachable? check the startup logs.

**Reply by JamieWuJinYi** [reply]
Thanks @danny-avila 
Yes,  rag api reachable
Now it times out and then an error is reported, but after a while the file can be found in my attachment, indicating that it has actually been uploaded successfully.
You can look at the first picture mentioned above.

Can the timeout of rag_api be set? Can the timeout of rag_api be set? It would be best if it is the same as librechat api.

**Reply by whindsx** [reply]
> Can the timeout of rag_api be set? Can the timeout of rag_api be set? It would be best if it is the same as librechat api.

Setting/extending 'proxy_read_timeout' in client/nginx.conf worked for me.

**Comment by jcbartle** [comment]
Good evening, folks. I'm throwing out a call for help on an issue I believe is related.

We have LibreChat and RAG API deployed within Azure App Services. Uploading smaller files works just fine. I'm running into issues when trying to upload a 35 MB HTML file (equates to ~ 3,800 pages of text). Based on the RAG API logs, this is getting broken up into ~ 110 chunks. These embedding calls all succeed; however, they take about 9 minutes to complete, start to finish.

The problem I have is that the LibreChat interface is timing out consistently after ~ 240 seconds with DevTools showing either an HTTP 499 or 504 response. After adjusting various timeout settings (`instance.defaults.timeout = 600000` in `axios.ts`; both `proxy_read_timeout` and `proxy_send_timeout` to `600s` in `nginx.conf`) with no luck, I discovered [this limitation](https://learn.microsoft.com/en-us/troubleshoot/azure/app-service/web-apps-performance-faqs#why-does-my-request-time-out-after-230-seconds) ("feature") of Azure App Service:

>Azure Load Balancer has a default idle timeout setting of four minutes. This setting is generally a reasonable response time limit for a web request. Therefore, App Service returns a timeout to the client if your application does not return a response within approximately 240 seconds (230 seconds on Windows app, 240 seconds on Linux app).

@danny-avila - what do you think about this issue? I see two paths forward. One possibility is adding an embedding status check to RAG API which could be periodically queried by the LibreChat client. This would likely follow the standard pattern of sending back an HTTP 202 upon submitting the request, with a periodic polling interval to get the status of the request. I actually raised this a while back in [this RAG API discussion](https://github.com/danny-avila/rag_api/discussions/182). I realize this would require changes on both the RAG API implementation and LibreChat. The other option which wouldn't require adjusting LibreChat would be to have a setting within RAG API for concurrent embeddings. Right now, they appear to be sequential. If we could bump this up to making 10 concurrent requests (or some other reasonable value), that would likely bring the time-to-completion well under the four-minute mark.
